### PR TITLE
Support i4 types in ConstEval

### LINKED ISSUE #15681: Support i4 types in ConstEval
In data-tiling on i4 matmul cases, we want to be able to const-eval for packing i4 constants.

### PR DESCRIPTION
**Pull request by Max191**
Fixes https://github.com/openxla/iree/issues/15681

### CODE REVIEW

**Comment by Max191** [comment]
This needs to wait until https://github.com/openxla/iree/pull/15558 and https://github.com/llvm/llvm-project/pull/73174 have landed.

**Comment by cc10512** [comment]
Can someone please comment on the status of this PR?

**Comment by stellaraccident** [comment]
> Can someone please comment on the status of this PR?

We probably lost track of finishing this and should land it. It's been a bit of a long term project to get all of the sub-byte stuff all the way. The linked patch was the next pre-requsisite needed to write meaningful tests, so I think this (or something like it) could go forward. In general, i4/i2/i1 would be good to support "natively". Outside of that, there is no common ground and we are tending towards solutions that don't require anyone to agree on the layout.

In our work, we've mostly switched to bit-casting the sub-byte or structured element type tensors from a higher bit depth tensor (sometimes structured and sometimes just 1d of i8). Still want to finish this, but bitcast unlocks a lot more, and it gets around the problem of needing to choose one packed layout for weird cases at rest. The thing that generates the program knows how it is physically manifested and can just cast it to the logical view on load, and that is a lot easier on the tools.

I expect that Max authored this patch at a point when i4 globals were the only way we were operating. But then about this time, we switched to Turbine models, and they use the above approach, happening to skate around the limitations. Like I said, would be good to finish, but that is probably why it stalled.

**Comment by Max191** [comment]
Yes, as Stella said, this PR became less of a priority since I opened it, so it has not been ready to land. The two referenced patches were needed but the reason I never finished this is that there is another missing piece still to enable i4 const eval. To evaluate i4 tensors, they are bitcasted before hoisting them into globals, but I still don't think the same bitcasting trick has been implemented for singular i4 constant values (non-tensors). Enabling i4 types in const eval was getting the job done for the use case we had because there was an i4 padding value constant that didn't need to be evaluated. However, to land this properly, we need to support the actual evaluation of i4 values through some bitcast trick like we do for i4 tensors. I've been meaning to come back and do this, but I haven't had cycles for a little while

**Comment by dcaballe** [comment]
Thanks for clarifying! @Max191, if you could send some pointers to where the work is needed we can try to help if you don't have cycles right now, unless it's easier for you to fix it yourself. ConstEval on i4 is high priority for us right now.

I'm also wondering if we could do an incremental enablement, esp. if i4 tensors are already working. Then we can work on supporting the scalar cases.

**Comment by qedawkins** [comment]
> Thanks for clarifying! @Max191, if you could send some pointers to where the work is needed we can try to help if you don't have cycles right now, unless it's easier for you to fix it yourself. ConstEval on i4 is high priority for us right now.
> 
> I'm also wondering if we could do an incremental enablement, esp. if i4 tensors are already working. Then we can work on supporting the scalar cases.

Could help to understand exactly the case you need to support, but essentially we just need to insert some casts to get around ABI issues (we can't pass around single i4 values, and that's doubly true once we get to weird struct data types).

For tensor<*xi4>, this should work from the perspective of consteval/hoist_into_globals based on these examples: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/GlobalOptimization/test/flow_hoist_into_globals.mlir.

For i4 scalar values, we can add additional type interfaces here: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp

Although the interface registration for scalar types can probably happen in Util given that it would just be `arith` based bitwise arithmetic and there would be no dependence on Flow/relevance to GlobalOptimization directly.

**Comment by Max191** [comment]
> Thanks for clarifying! @Max191, if you could send some pointers to where the work is needed we can try to help if you don't have cycles right now, unless it's easier for you to fix it yourself. ConstEval on i4 is high priority for us right now.
> 
> I'm also wondering if we could do an incremental enablement, esp. if i4 tensors are already working. Then we can work on supporting the scalar cases.

Thanks @qedawkins for providing some pointers. To add some additional context, const eval on i4 tensors already works, but the i4 scalars are needed when you do data tiling (and therefore packing) because of i4 padding values. Technically, this PR should be enough to test const eval on most normal data tiling workloads because the i4 padding value does not need to be evaluated and stored as a hoisted global. But to get this into a landable state I don't think there is too much that needs to be done.

As Quinn mentioned, https://github.com/openxla/iree/pull/15558 adds an interface that handles i4 tensor, which is used here: https://github.com/openxla/iree/blob/aa3d9099a0142dbb8df829d09e6cd5c137bee0db/compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp#L123-L128
And also here: https://github.com/openxla/iree/blob/aa3d9099a0142dbb8df829d09e6cd5c137bee0db/compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp#L187-L192

The basic idea of that interface is that it takes the tensor type and creates a `flow.tensor.bitcast` to cast it into `i8` (or some byte aligned type) just before the constant is hoisted. Then, it is bitcasted back into `i4` after the hoisted global is loaded. This same logic should work for scalars, but it just needs to be supported (perhaps by adding an additional type interface like Quinn suggests).

Outside of extending the hoisting to scalars, I think the changes in this PR should suffice to flip the switch and enable const eval on `i4`

**Comment by benvanik** [comment]
It'd be good to ground this with an example. We shouldn't need to do anything special for scalar types outside of some VM-level things - everything above VM level should be fine with any scalar types. Const-eval, hoisting, etc should be completely orthogonal to scalar types being discussed and I want to make sure we aren't adding special handling we don't need.

(not saying there's no work to be done, just that all the way up until conversion to vm there should be nothing needed)

File: compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/OpOracle.cpp [hunk 1]
```diff
@@ -61,8 +61,8 @@ bool isLegalConstExprType(Type t) {
     // support, but for now the consteval JIT has interop limitations. Lift
     // this restriction when the JIT interops for all types.
     auto bitWidth = t.getIntOrFloatBitWidth();
-    return bitWidth == 1 || bitWidth == 8 || bitWidth == 16 || bitWidth == 32 ||
-           bitWidth == 64;
+    return bitWidth == 1 || bitWidth == 4 || bitWidth == 8 || bitWidth == 16 ||
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/OpOracle.cpp:64]: I'd not want to special case 4 here as it's really any power-of-two value up to 64 bits (so 2 should be added, at which point we can do `isPowerOf2(bitWidth) && bitWidth <= 64`

**Comment by bviyer** [comment]
hi @Max191,  I took this patch and added a change to get it working for scalar i4s. I hope it is OK with you. I have added you as a co-author to this patch. (https://github.com/openxla/iree/pull/16321)

**Pull request closed without merging**